File: client/src/lib/shelf.ts
export type ShelfWordLike = {
  word?: { text?: string | null } | null;
  song?: { id?: string | null } | null;
};

/** One shelf card per word (case-insensitive). Keeps the first / newest row. */
export function uniqueShelfWords<T extends ShelfWordLike>(items: T[]): T[] {
  const seen = new Set<string>();
  const out: T[] = [];
  for (const item of items || []) {
    const key = String(item?.word?.text || '').toLowerCase().trim();
    if (!key || seen.has(key)) continue;
    seen.add(key);
    out.push(item);
  }
  return out;
}

File: mobile/lib/utils/shelf.dart
List<Map<String, dynamic>> uniqueShelfWords(List<Map<String, dynamic>> items) {
  final seen = <String>{};
  final out = <Map<String, dynamic>>[];
  for (final item in items) {
    final rawWord = item['word'];
    final text = rawWord is Map
        ? (rawWord['text'] ?? '').toString()
        : (rawWord ?? item['text'] ?? '').toString();
    final key = text.toLowerCase().trim();
    if (key.isEmpty || seen.contains(key)) continue;
    seen.add(key);
    out.add(item);
  }
  return out;
}
